sdk/projects/azure-projects/azure/projects/_resource.py

```diff
@@ -0,0 +1,632 @@
+            self._settings["subscription"] = self.parent._settings["subscription"]
+        elif "resource_group" in self.properties:
+            self._settings["resource_group"] = self.properties["resource_group"]._settings["name"]
+            self._settings["subscription"] = self.properties["resource_group"]._settings["subscription"]
+        else:
+            self._settings["resource_group"] = StoredPrioritizedSetting(
+                name="resource_group",
+                env_vars=_build_envs(self._prefixes, ["RESOURCE_GROUP"]),
+                suffix=self._env_suffix,
+            )
+            self._settings["subscription"] = StoredPrioritizedSetting(
+                name="subscription",
+                env_var="AZURE_SUBSCRIPTION_ID",
+                suffix=self._env_suffix,
+                default=self.properties.get("subscription", _unset),
+            )
+        if kwargs:
+            raise TypeError(f"Resource {self.__class__.__name__} got unexpected kwargs: {list(kwargs.keys())}")
+
+    def __repr__(self) -> str:
+        name = f'\'{self.properties["name"]}\'' if "name" in self.properties else "<default>"
+        return f"{self.__class__.__name__}({name})"
+
+    def __eq__(self, value: Any) -> bool:
+        # Resources are considered equal if they (and their parents) have the same type and same name.
+        try:
+            parent_eq = True
+            if self.parent:
+                parent_eq = self.parent == value.parent
+            return (
+                parent_eq
+                and value.resource == self.resource
+                and value.properties.get("name") == self.properties.get("name")
+            )
+        except Exception:  # pylint: disable=broad-exception-caught
+            return False
+
+    @property
+    def resource(self) -> str:
+        raise NotImplementedError("This must be implemented by child resources.")
+
+    @property
+    def version(self) -> str:
+        # TODO: Support version override
+        raise NotImplementedError("This must be implemented by child resources.")
+
+    @classmethod
+    def reference(
+        cls,
+        *,
+        name: Optional[Union[str, Parameter]] = None,
+        resource_group: Optional[Union[str, Parameter, "ResourceGroup[ResourceReference]"]] = None,
+        subscription: Optional[Union[str, Parameter]] = None,
+        parent: Optional[Resource] = None,
+    ) -> Self:
+        if parent and resource_group:
+            raise ValueError("Cannot specify both parent and resource_group.")
+        properties = ResourceReference(name=name) if name else {}
+        if resource_group:
+            if isinstance(resource_group, (str, Parameter)):
+                from .resources.resourcegroup import ResourceGroup
+
+                resource_group = ResourceGroup.reference(name=resource_group, subscription=subscription)
+            properties["resource_group"] = resource_group
+        elif subscription:
```

Comment:
Feels like resource_group and subscription are mutually exclusive, should this be explictly tested and raise an exception?

Reply:
No they're not - as you can see on line 251 that if `resource_group` is provided, `subscription` is passed in there, whereas if _only_ `subscription` is provided then we handle that separately.